sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java

```diff
@@ -0,0 +1,465 @@
+        builder.add(DisplayData.item("dataSource", dataSource.getClass().getName()));
+        builder.add(DisplayData.item("query", query));
+        builder.addIfNotNull(DisplayData.item("username", username));
+      }
+
+      public DataSource getDataSource() {
+        return dataSource;
+      }
+
+      public String getQuery() {
+        return query;
+      }
+
+      @Nullable
+      public String getUsername() {
+        return username;
+      }
+
+      @Nullable
+      public String getPassword() {
+        return password;
+      }
+    }
+
+    /**
+     * A {@link DoFn} executing the SQL query to read from the database.
+     */
+    public static class ReadFn<T> extends DoFn<JdbcOptions, T> {
+
+      private final RowMapper<T> rowMapper;
+
+      private ReadFn(RowMapper<T> rowMapper) {
+        this.rowMapper = rowMapper;
+      }
+
+      @ProcessElement
+      public void processElement(ProcessContext context) throws Exception {
+        JdbcOptions options = context.element();
+
+        try (Connection connection = (options.getUsername() != null)
+            ? options.getDataSource().getConnection(options.getUsername(), options.getPassword())
+            : options.getDataSource().getConnection()) {
+
+          try (PreparedStatement statement = connection.prepareStatement(options.getQuery())) {
+            try (ResultSet resultSet = statement.executeQuery()) {
+              while (resultSet.next()) {
+                T record = rowMapper.mapRow(resultSet);
+                  context.output(record);
+              }
+            }
+          }
+        }
+      }
+    }
+  }
+
+  /**
+   * An interface used by the JdbcIO Write to set the parameters of the {@link PreparedStatement}
+   * used to setParameters into the database.
+   */
+  public interface PreparedStatementSetter<T> extends Serializable {
+    void setParameters(T element, PreparedStatement preparedStatement) throws Exception;
+  }
+
+  /**
+   * A {@link PTransform} to write to a JDBC datasource.
+   */
+  public static class Write<T> extends PTransform<PCollection<T>, PDone> {
+
+    public Write<T> withDataSource(DataSource dataSource) {
+      return new Write<>(writer.withDataSource(dataSource));
+    }
+
+    public Write<T> withQuery(String query) {
+      return new Write<>(writer.withQuery(query));
+    }
+
+    public Write<T> withUsername(String username) {
+      return new Write<>(writer.withUsername(username));
+    }
+
+    public Write<T> withPassword(String password) {
+      return new Write<>(writer.withPassword(password));
+    }
+
+    public <X> Write<X> withPreparedStatementSetter(
+        PreparedStatementSetter<X> preparedStatementSetter) {
+      return new Write<>(writer.withPreparedStatementSetter(preparedStatementSetter));
+    }
+
+    public Write<T> withBatchSize(long batchSize) {
+      return new Write<>(writer.withBatchSize(batchSize));
+    }
+
+    private final JdbcWriter writer;
```

Comment:
The read function is called ReadFn, but this one is called Writer. Please eliminate the inconsistency. The more common style is BlahFn.